File: apps/host/internal/githubapi/client_test.go
package githubapi

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strconv"
	"sync/atomic"
	"testing"
	"time"

	pb "armadra.local/host/gen/armadra/v1"
)

// Every test here talks to a local httptest server. Nothing in this package's
// suite ever reaches api.github.com.

var testContext = context.Background()

func testClient(t *testing.T, handler http.Handler) (*Client, *httptest.Server) {
	t.Helper()
	server := httptest.NewServer(handler)
	t.Cleanup(server.Close)
	client, err := New(Options{
		APIBase: server.URL,
		Token:   func(context.Context) (string, error) { return "test-token-not-real", nil },
		HTTP:    server.Client(),
		Now:     time.Now,
		Sleep:   func(context.Context, time.Duration) error { return nil },
	})
	if err != nil {
		// The mock is http, which NormalizeAPIBase rightly refuses, so the test
		// builds the client directly against the parsed mock URL instead.
		parsed, parseErr := url.Parse(server.URL)
		if parseErr != nil {
			t.Fatal(parseErr)
		}
		client = &Client{
			base:     parsed,
			token:    func(context.Context) (string, error) { return "test-token-not-real", nil },
			http:     server.Client(),
			now:      time.Now,
			sleep:    func(context.Context, time.Duration) error { return nil },
			agent:    "armadra-host-test",
			attempts: 3,
			cache:    map[string]*cacheEntry{},
		}
		client.http.CheckRedirect = func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }
	}
	return client, server
}

// A poll loop has to stay inside the quota, so a matching ETag must replay the
// stored body rather than counting as a fresh read.
func TestConditionalRequestsReplayTheStoredBody(t *testing.T) {
	var requests atomic.Int64
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requests.Add(1)
		w.Header().Set("ETag", `W/"abc"`)
		w.Header().Set("X-RateLimit-Limit", "5000")
		w.Header().Set("X-RateLimit-Remaining", "4999")
		if r.Header.Get("If-None-Match") == `W/"abc"` {
			w.WriteHeader(http.StatusNotModified)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`[{"number":1,"title":"first"}]`))
	}))
	first, err := client.Get(testContext, "/repos/o/r/issues", nil)
	if err != nil || first.FromCache || len(first.Body) == 0 {
		t.Fatalf("first read: %v %+v", err, first)
	}
	second, err := client.Get(testContext, "/repos/o/r/issues", nil)
	if err != nil {
		t.Fatal(err)
	}
	if !second.FromCache || string(second.Body) != string(first.Body) {
		t.Fatal("a 304 must replay the stored body and say it came from cache")
	}
	if second.Rate.Limit != 5000 {
		t.Fatal("rate limit headers must survive a 304")
	}
	if requests.Load() != 2 {
		t.Fatalf("expected two requests, got %d", requests.Load())
	}
}

// A write that reached the remote must never be sent twice: a duplicate comment
// or a second merge is worse than reporting an unknown outcome.
func TestWritesAreNeverRetriedAndReportAnUnknownOutcome(t *testing.T) {
	var attempts atomic.Int64
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		attempts.Add(1)
		w.WriteHeader(http.StatusInternalServerError)
	}))
	_, err := client.Write(testContext, http.MethodPost, "/repos/o/r/issues", map[string]any{"title": "x"})
	if CodeOf(err) != CodeUnknownOutcome {
		t.Fatalf("a 500 on a write must be an unknown outcome, got %v", err)
	}
	if attempts.Load() != 1 {
		t.Fatalf("the write was attempted %d times", attempts.Load())
	}
	// A read, which has no side effect, is allowed to retry.
	attempts.Store(0)
	if _, err = client.Get(testContext, "/repos/o/r/issues", nil); CodeOf(err) != CodeUnavailable {
		t.Fatalf("a failing read reported %v", err)
	}
	if attempts.Load() != 3 {
		t.Fatalf("the read was attempted %d times", attempts.Load())
	}
}

// GitHub answers 403 both for "you may not" and for its secondary rate limit.
// Merging them would tell a user to fix a permission when they should wait.
func TestSecondaryRateLimitIsDistinguishedFromPermission(t *testing.T) {
	limited, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Retry-After", "1")
		w.WriteHeader(http.StatusForbidden)
	}))
	if _, err := limited.Get(testContext, "/repos/o/r/issues", nil); CodeOf(err) != CodeRateLimited {
		t.Fatalf("a throttled 403 reported %v", err)
	}
	forbidden, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-RateLimit-Limit", "5000")
		w.Header().Set("X-RateLimit-Remaining", "4000")
		w.WriteHeader(http.StatusForbidden)
	}))
	if _, err := forbidden.Get(testContext, "/repos/o/r/issues", nil); CodeOf(err) != CodePermission {
		t.Fatalf("an unthrottled 403 reported %v", err)
	}
}

// A redirect would let the remote move a request, and its bearer token, to
// another authority.
func TestRedirectsAreRefusedRatherThanFollowed(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "https://evil.example.com/repos/o/r", http.StatusFound)
	}))
	if _, err := client.Get(testContext, "/repos/o/r", nil); CodeOf(err) != CodeInvalid {
		t.Fatalf("a redirect reported %v", err)
	}
}

// A path is a path. Nothing a caller passes may re-target the request.
func TestRequestPathsCannotLeaveTheConfiguredBase(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}))
	for _, path := range []string{"repos/o/r", "//evil.example.com/x", "/repos/o/r?x=1", "/repos/o/r#f", "/repos/o/\x00r"} {
		if _, err := client.Get(testContext, path, nil); CodeOf(err) != CodeInvalid {
			t.Fatalf("path %q was accepted", path)
		}
	}
}

// A caller cannot address a repository through a client built for another base.
func TestRepositoryReferencesMustMatchTheClientBase(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}))
	elsewhere := &pb.GithubRepositoryRef{Owner: "o", Name: "r", ApiBase: "https://ghe.example.com/api/v3"}
	if _, _, err := client.Repository(testContext, elsewhere, 1); CodeOf(err) != CodeInvalid {
		t.Fatalf("a foreign API base was accepted: %v", err)
	}
	bad := &pb.GithubRepositoryRef{Owner: "../..", Name: "r"}
	if _, _, err := client.Repository(testContext, bad, 1); CodeOf(err) != CodeInvalid {
		t.Fatalf("an invalid owner was accepted: %v", err)
	}
}

// The Issues endpoint returns pull requests too. Showing them in an Issues list
// would put a Close button on something where closing means another thing.
func TestPullRequestsAreFilteredOutOfIssueListings(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`[
          {"number":1,"title":"issue","state":"open","updated_at":"2026-09-05T10:00:00Z"},
          {"number":2,"title":"pull","state":"open","pull_request":{"url":"x"}}
        ]`))
	}))
	ref := &pb.GithubRepositoryRef{Owner: "o", Name: "r"}
	issues, _, err := client.Issues(testContext, ref, nil, 1, 50, 42)
	if err != nil {
		t.Fatal(err)
	}
	if len(issues) != 1 || issues[0].Number != 1 {
		t.Fatalf("listing returned %d entries", len(issues))
	}
	if issues[0].ObservedAtUnixMs != 42 {
		t.Fatal("an issue must carry when it was actually observed")
	}
	if issues[0].UpdatedAtUnixMs == 0 {
		t.Fatal("updated_at must survive; it is the read-before-write token")
	}
}

// Checks come from two independent surfaces. Reporting only one would show an
// empty, falsely reassuring result for a repository that uses the other.
func TestChecksCombineCheckRunsAndCommitStatuses(t *testing.T) {
	sha := "9fceb02d0ae598e95dc970b74767f19372d61af8"
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch {
		case r.URL.Path == "/repos/o/r/commits/"+sha+"/check-runs":
			_, _ = w.Write([]byte(`{"check_runs":[
              {"name":"build","status":"completed","conclusion":"success","app":{"name":"GitHub Actions","slug":"github-actions"},"details_url":"https://example.com/actions/runs/77/job/1"},
              {"name":"slow","status":"in_progress","app":{"name":"Other","slug":"other"}}
            ]}`))
		case r.URL.Path == "/repos/o/r/commits/"+sha+"/status":
			_, _ = w.Write([]byte(`{"statuses":[{"context":"legacy","state":"failure"}]}`))
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	ref := &pb.GithubRepositoryRef{Owner: "o", Name: "r"}
	summary, err := client.Checks(testContext, ref, sha, 7)
	if err != nil {
		t.Fatal(err)
	}
	if len(summary.Runs) != 3 || summary.HeadSha != sha {
		t.Fatalf("summary has %d runs for %s", len(summary.Runs), summary.HeadSha)
	}
	// A failure outranks everything, including a run still in progress.
	if summary.Rollup != pb.GithubCheckConclusion_GITHUB_CHECK_CONCLUSION_FAILURE {
		t.Fatalf("rollup was %v", summary.Rollup)
	}
	// Only a workflow run can actually be restarted; nothing else may claim it.
	if !summary.Runs[0].Rerunnable || summary.Runs[0].WorkflowRunId != 77 {
		t.Fatal("an Actions run must expose its workflow run")
	}
	if summary.Runs[1].Rerunnable || summary.Runs[2].Rerunnable {
		t.Fatal("a check with no rerun endpoint must not be offered one")
	}
	if summary.Runs[1].Conclusion != pb.GithubCheckConclusion_GITHUB_CHECK_CONCLUSION_PENDING {
		t.Fatal("an unfinished run is pending, never a pass")
	}
}

// A short SHA would let a merge name something other than exactly the head the
// reader saw.
func TestMergeRequiresAFullObjectName(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"merged":true,"sha":"1111111111111111111111111111111111111111"}`))
	}))
	ref := &pb.GithubRepositoryRef{Owner: "o", Name: "r"}
	if _, err := client.Merge(testContext, ref, 1, "9fceb02", "squash", "", ""); CodeOf(err) != CodeInvalid {
		t.Fatal("a short SHA was accepted for a merge")
	}
	sha, err := client.Merge(testContext, ref, 1, "9fceb02d0ae598e95dc970b74767f19372d61af8", "squash", "", "")
	if err != nil || sha != "1111111111111111111111111111111111111111" {
		t.Fatalf("merge returned %q (%v)", sha, err)
	}
}

// A merge the remote reports as not merged is a refusal, not a success.
func TestMergeRefusalIsNeverReportedAsSuccess(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"merged":false,"message":"not mergeable"}`))
	}))
	ref := &pb.GithubRepositoryRef{Owner: "o", Name: "r"}
	_, err := client.Merge(testContext, ref, 1, "9fceb02d0ae598e95dc970b74767f19372d61af8", "squash", "", "")
	if CodeOf(err) != CodeConflict {
		t.Fatalf("an unmerged response reported %v", err)
	}
}

// A rejected token has one repair, and it is not "wait" or "reload".
func TestTokenRejectionIsItsOwnCode(t *testing.T) {
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
	}))
	var problem *Error
	_, err := client.Get(testContext, "/user", nil)
	if !errors.As(err, &problem) || problem.Code != CodeUnauthenticated || problem.Reason != "TOKEN_REJECTED" {
		t.Fatalf("a 401 reported %v", err)
	}
}

// The Host must never send an anonymous request that silently reads only what
// is public.
func TestAnEmptyTokenIsRefusedBeforeTheRequestLeaves(t *testing.T) {
	var requests atomic.Int64
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { requests.Add(1) }))
	client.token = func(context.Context) (string, error) { return "", nil }
	if _, err := client.Get(testContext, "/user", nil); CodeOf(err) != CodeUnauthenticated {
		t.Fatal("an empty token was not refused")
	}
	if requests.Load() != 0 {
		t.Fatal("a request left the process without a credential")
	}
}

// The Host holds the token for the length of a call, and asks again next time,
// so revoking a credential stops working immediately.
func TestEveryRequestAsksTheTokenSourceAgain(t *testing.T) {
	var asked atomic.Int64
	client, _ := testClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}))
	client.token = func(context.Context) (string, error) {
		return "token-" + strconv.FormatInt(asked.Add(1), 10), nil
	}
	for range 3 {
		if _, err := client.Get(testContext, "/user", nil); err != nil {
			t.Fatal(err)
		}
	}
	if asked.Load() != 3 {
		t.Fatalf("the token source was consulted %d times", asked.Load())
	}
}

File: apps/desktop/src/core/github/client.test.ts
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { MAX_CACHED_ISSUES, MAX_CACHE_BYTES, nextPage, perPage } from "./client";
import { codeOf } from "./errors";
import { fakeGithub, type FakeGithub } from "./fixture";

/** 移植自 `apps/host/internal/githubapi/client_test.go`。 */
describe("GitHub 传输层", () => {
  let github: FakeGithub;

  beforeEach(async () => {
    github = await fakeGithub();
  });

  afterEach(async () => {
    await github.close();
  });

  it("ETag 命中时放回存着的正文，并标成来自缓存", async () => {
    github.route("GET /repos/a/b", {
      body: { full_name: "a/b" },
      etag: '"v1"',
    });
    const client = github.client();
    const first = await client.get("/repos/a/b");
    expect(first.fromCache).toBe(false);
    const second = await client.get("/repos/a/b");
    expect(second.fromCache).toBe(true);
    expect(second.status).toBe(200);
    expect(JSON.parse(second.body.toString("utf8"))).toEqual({
      full_name: "a/b",
    });
    // 第二次确实发出去了，只是远端答了 304——省的是配额不是往返。
    expect(github.requests).toHaveLength(2);
    expect(github.requests[1]?.headers["if-none-match"]).toBe('"v1"');
  });

  it("只从 Link 头里读页码，从不交回远端 URL", () => {
    expect(
      nextPage('<https://api.github.com/repos/a/b/issues?page=3>; rel="next"'),
    ).toBe(3);
    // 一个 URL 形式的游标会让响应操纵下一次请求，所以只有页码被取出来。
    expect(nextPage('<https://evil.example/x?page=2>; rel="prev"')).toBe(0);
    expect(nextPage("")).toBe(0);
    expect(nextPage('<https://api.github.com/x?page=99999>; rel="next"')).toBe(0);
  });

  it("读失败会退避重试，写一次都不重试", async () => {
    github.once("GET /repos/a/b", { status: 503 });
    github.route("GET /repos/a/b", { body: { full_name: "a/b" } });
    const client = github.client();
    await expect(client.get("/repos/a/b")).resolves.toBeDefined();
    expect(github.requests.filter((r) => r.method === "GET")).toHaveLength(2);

    github.route("POST /repos/a/b/issues", { status: 503 });
    // 500 段的写报 UNKNOWN_OUTCOME：请求可能已经生效，重试会重复一条评论。
    const error = await client
      .write("POST", "/repos/a/b/issues", { title: "x" })
      .catch((value: unknown) => value);
    expect(codeOf(error)).toBe("UNKNOWN_OUTCOME");
    expect(
      github.requests.filter((r) => r.method === "POST"),
    ).toHaveLength(1);
  });

  it("403 只有在配额头说被限流时才读成限流", async () => {
    const client = github.client();
    github.once("GET /repos/a/b", {
      status: 403,
      headers: { "x-ratelimit-remaining": "17", "x-ratelimit-limit": "60" },
    });
    expect(codeOf(await client.get("/repos/a/b").catch((e: unknown) => e))).toBe(
      "PERMISSION_DENIED",
    );
    github.once("GET /repos/a/c", {
      status: 403,
      headers: { "x-ratelimit-remaining": "0", "x-ratelimit-limit": "60" },
    });
    expect(codeOf(await client.get("/repos/a/c").catch((e: unknown) => e))).toBe(
      "RESOURCE_EXHAUSTED",
    );
  });

  it("重定向被当成错误，而不是跟着走到另一个 authority", async () => {
    github.route("GET /repos/a/b", {
      status: 302,
      headers: { location: "https://evil.example/repos/a/b" },
    });
    const client = github.client({ attempts: 1 });
    const error = await client.get("/repos/a/b").catch((value: unknown) => value);
    expect(codeOf(error)).toBe("INVALID_ARGUMENT");
    // 只联系了配置好的那个 base。
    expect(github.requests).toHaveLength(1);
  });

  it("会改写目的地的路径一律拒绝", async () => {
    const client = github.client();
    for (const path of ["repos/a/b", "/repos//a", "/repos/a?x=1", "/a\\b"]) {
      const error = await client.get(path).catch((value: unknown) => value);
      expect(codeOf(error)).toBe("INVALID_ARGUMENT");
    }
    expect(github.requests).toHaveLength(0);
  });

  it("限流头会让下一次写在本地被拒，而不是加深次级限流", async () => {
    const client = github.client();
    github.once("GET /repos/a/b", {
      status: 429,
      headers: { "retry-after": "30" },
    });
    await client.get("/repos/a/b").catch(() => undefined);
    const error = await client
      .write("POST", "/repos/a/b/issues", {})
      .catch((value: unknown) => value);
    expect(codeOf(error)).toBe("RESOURCE_EXHAUSTED");
    expect(github.requests.filter((r) => r.method === "POST")).toHaveLength(0);
  });

  it("X-OAuth-Scopes 报出来的 scope 被读出来，没报就是空列表", async () => {
    const client = github.client();
    github.once("GET /user", {
      body: { login: "octocat" },
      headers: { "x-oauth-scopes": "repo, read:org" },
    });
    expect((await client.get("/user")).oauthScopes).toEqual([
      "repo",
      "read:org",
    ]);
    github.once("GET /user", { body: { login: "octocat" } });
    // 细粒度令牌一个都不报；空列表的意思是「没报」，不是「没有权限」。
    expect((await client.get("/user")).oauthScopes).toEqual([]);
  });

  it("缓存满了逐出最近最少使用的那条，而不是继续长", async () => {
    const client = github.client();
    for (let index = 0; index < 120; index += 1) {
      github.route(`GET /repos/a/r${index}`, {
        body: { id: index },
        etag: `"e${index}"`,
      });
      await client.get(`/repos/a/r${index}`);
    }
    // 第 0 条早被逐出了：再读一次不会命中 304。
    const before = github.requests.length;
    const again = await client.get("/repos/a/r0");
    expect(again.fromCache).toBe(false);
    expect(github.requests.length).toBe(before + 1);
    expect(client.cachedBytes()).toBeLessThanOrEqual(MAX_CACHE_BYTES);
  });

  it("缓存的条目上限由「最多 10000 条 issue」导出", () => {
    expect(MAX_CACHED_ISSUES).toBe(10_000);
    expect(MAX_CACHE_BYTES).toBe(64 * 1024 * 1024);
  });

  it("per_page 被夹在 GitHub 自己的上限内", () => {
    expect(perPage(0)).toBe("100");
    expect(perPage(30)).toBe("30");
    expect(perPage(5_000)).toBe("100");
  });
});
